<a href="https://colab.research.google.com/github/sarjakpatel/AIRoyal/blob/master/IrisMLRoyal.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [9]:
import numpy as np
import pandas as pd
from keras.utils import to_categorical


In [10]:
dataset=pd.read_csv('/content/Iris.csv')
dataset

Id  SepalLengthCm  ...  PetalWidthCm         Species
0      1            5.1  ...           0.2     Iris-setosa
1      2            4.9  ...           0.2     Iris-setosa
2      3            4.7  ...           0.2     Iris-setosa
3      4            4.6  ...           0.2     Iris-setosa
4      5            5.0  ...           0.2     Iris-setosa
..   ...            ...  ...           ...             ...
145  146            6.7  ...           2.3  Iris-virginica
146  147            6.3  ...           1.9  Iris-virginica
147  148            6.5  ...           2.0  Iris-virginica
148  149            6.2  ...           2.3  Iris-virginica
149  150            5.9  ...           1.8  Iris-virginica

[150 rows x 6 columns]

In [11]:
x=dataset.iloc[:,1:5].values
y=dataset.iloc[:,5].values

from sklearn.preprocessing import LabelEncoder
encoder=LabelEncoder()
y1=encoder.fit_transform(y)
#print(y1)


y2=pd.get_dummies(y1).values
#print(y2)

In [12]:
from sklearn.model_selection import train_test_split


x_train,x_test,y2_train,y2_test = train_test_split(x,y2,test_size=0.2,random_state=0)
#print(x_test)

In [32]:
from keras.models import Sequential
from keras.layers import Dense
from keras.optimizers import SGD, Adam

model = Sequential()

model.add(Dense(10, input_shape=(4,), activation='tanh'))
model.add(Dense(8, activation='tanh'))
model.add(Dense(6, activation='tanh'))
model.add(Dense(3, activation='softmax'))


model.compile(Adam(lr=0.04), 'categorical_crossentropy', metrics=['accuracy'])

model.summary()


Model: "sequential_10"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_46 (Dense)             (None, 10)                50        
_________________________________________________________________
dense_47 (Dense)             (None, 8)                 88        
_________________________________________________________________
dense_48 (Dense)             (None, 6)                 54        
_________________________________________________________________
dense_49 (Dense)             (None, 3)                 21        
Total params: 213
Trainable params: 213
Non-trainable params: 0
_________________________________________________________________


In [33]:
model.fit(x_train, y2_train, epochs = 100)
y_pred=model.predict(x_test)



Epoch 1/100
4/4 [==============================] - 0s 3ms/step - loss: 1.0566 - accuracy: 0.5000
Epoch 2/100
4/4 [==============================] - 0s 2ms/step - loss: 0.8052 - accuracy: 0.6500
Epoch 3/100
4/4 [==============================] - 0s 2ms/step - loss: 0.5162 - accuracy: 0.8333
Epoch 4/100
4/4 [==============================] - 0s 2ms/step - loss: 0.3529 - accuracy: 0.9083
Epoch 5/100
4/4 [==============================] - 0s 2ms/step - loss: 0.2420 - accuracy: 0.9333
Epoch 6/100
4/4 [==============================] - 0s 2ms/step - loss: 0.1776 - accuracy: 0.9417
Epoch 7/100
4/4 [==============================] - 0s 2ms/step - loss: 0.1635 - accuracy: 0.9667
Epoch 8/100
4/4 [==============================] - 0s 4ms/step - loss: 0.0986 - accuracy: 0.9750
Epoch 9/100
4/4 [==============================] - 0s 2ms/step - loss: 0.1149 - accuracy: 0.9583
Epoch 10/100
4/4 [==============================] - 0s 2ms/step - loss: 0.1805 - accuracy: 0.9333
Epoch 11/100
4/4 [===========

## **Conclusion**:

 Adam optimizer:

    Number of layer = 4 (10, 8, 6, 3)
    Learning rate = 0.04
    epoch=100
    loss: 0.0603 - accuracy: 0.9833

    Number of layer = 5 (10, 10, 8, 6, 3)
    Learning rate = 0.04
    epoch=100
    loss: 0.1749 - accuracy: 0.9500

    Number of layer = 6 (10, 10, 8, 8, 6, 3)
    Learning rate = 0.04
    epoch=100
    loss: 0.0842 - accuracy: 0.9750




  SGD:

    
    Number of layer = 4 (10, 8, 6, 3)
    Learning rate=0.04
    epoch=100
    loss: 0.1827 - accuracy: 0.9667

    Number of layer = 5 (10, 10, 8, 6, 3)
    Learning rate = 0.04
    epoch=100
    loss: 0.2072 - accuracy: 0.9667 

    Number of layer = 6 (10, 10, 8, 8, 6, 3)
    Learning rate = 0.04
    epoch=100
    loss: 0.1511 - accuracy: 0.9417



From the above analysis we can say that Adam is more efficient than SGD. Here, we can't see much difference because dataset is very small it contains only 150 rows. When dataset is very large then we can easily see the major difference between both the optimizers. In both the optimizers if we add more number of layers in the model, result is almost similar because this dataset has only 4 features and we can easily find the pattern between these features with the help of 4 layers only. So there is no need to add extra layer to the model for this dataset. If we add more layers then it takes more time to converge because the number of calculation and parameters are increased. Apart from this small dataset if dataset is very large and it has many number of features & if we add more number of layers in the model then it gives very good accuracy.
